### ISSUE TITLE
Design: Add files to IPFS

### ISSUE THREAD

**Issue by olizilla** [ipfs-shipyard/pm-ipfs-gui#35]
All three apps provide drag'n'drop and and a file picker dialog to add a file, but there are problems with each one. What is the ideal UI/UX for adding files to your IPFS repo.

This is closely related to the next step **Share Files via IPFS** discussed in #34 

The current state of things is described here: https://github.com/ipfs-shipyard/pm-ipfs-gui/blob/master/research/README.md#add-files

**Comment by olizilla** [comment]
The fundamentals... regardless of app, file-drop and open a native file picker dialog should be supported. The user should be able to add any file type or directory from the same action. 

Once the file is selected we need to show a progress bar to indicate that it we're working on it. The progress bar should be deterministic, showing what percentage of the file has been processed. The progress bar could be skipped if it will take less than 1s to add.

Once completed, the filename and CID should be shown, along with a buttons to share the file (See: #34).

**Comment by hacdias** [comment]
And those files should be added to MFS, right?

**Comment by akrych** [comment]
It should also be **one button** for adding new files or folders.

**Comment by hacdias** [comment]
@akrych unfortunately that's something I just *can't* do due to Operating System restrictions.

```
Note: On Windows and Linux an open dialog can not be both a file selector and a directory selector, so if you set properties to ['openFile', 'openDirectory'] on these platforms, a directory selector will be shown.
```

**Comment by hacdias** [comment]
Although, we could have one button that after clicking on it we could choose if we wanted to pick a file or a folder.


![image](https://user-images.githubusercontent.com/5447088/37556332-5530be66-29ec-11e8-989e-dee73fb799db.png)

**Comment by lidel** [comment]
I believe we have three ways of adding content to MFS (which implicitly pins it) right now:
- add a file
- add a directory (could be entire tree)
- add a NURI
  - this effectively replaces current pinning UI (based on `ipfs.pin`) with much simpler abstraction of MFS (`ipfs.files`) – i elaborated on value added by this approach in https://github.com/ipfs-shipyard/pm-ipfs-gui/issues/10#issuecomment-373928862
  - would be fantastic if pinning progress bar looked the same as for other options

**Comment by akrych** [comment]
@hacdias AAa ok :D Did't know that. Ok so we will do one button and after click choose if we wanted to pick a file or a folder. 

@lidel After consideration I like the simplicity of solution with add NURI as pinning - I just afraid about non-technical user - how we say them what "NURI" is. Still thinking about it

**Comment by olizilla** [comment]
Of note: `NURI` is dead. Long live `Path`

> NURI (Nested URI) is just a path. This is a deprecated term, but you'll see it around.

https://gateway.ipfs.io/ipns/ipfs-docs-research.robbrackett.com/html/Interviews/2018-02-02-Lars.html

**Comment by akrych** [comment]
@olizilla  So you suggest to call it "Add path" ? In my opinion still not so clear for non-technical user. 
In word "pin" we had some more explenation already. Just thinking thinking out loud :)

**Comment by olizilla** [comment]
@akrych keep those _out loud_ thoughts coming. I agree that `Path` just on it's own might not be enough. We should call it Path, or IPFS Path, to keep the terminology consistent with core, but we could use placeholder text or tooltips to show an example `/ipfs/QmHash`.

More generally, there are going to be a few concepts that will just have to be explained to every new user, so we should plan for an integrated help system from the start, along with a glossary of terms. @Mr0grog's research calls for a `concepts dictionary` too: https://github.com/ipfs/docs/issues/56

**Comment by olizilla** [comment]
I think we can fetch IPLD info about the data before fetching the whole file. I think we can make all this smarter by just fetching the metadata first, so that we can then offer the user more info like how big the file might be, and a chance to change their mind, along with a finite progress bar, that knows how many bytes to expect.

**Comment by olizilla** [comment]
we have to deal with the fact that we might not be able to find the Path that the user gives us if there are no copies currently being offered up.

**Comment by olizilla** [comment]
**What is someone trying to do when they add files to an IPFS gui?** Right now, I'm probably about to share those files someone.

Adding things to an IPFS gui today leaves with a duplicate of the file, in a repo that I can't easily see in my filesystem explorer, so it's like I'm adding a share-able copy of a file to a magic box. A bit like adding a file to your `Public` folder in OSX, and turning on local sharing, though I have no idea how successful that is as an idea. I once used that for LAN sharing but gave up.) As the repo structure is separate from the host file system, I can't easily incorporate IPFS into my personal filing system. I can't see a strong motive yet for people to add files to IPFS as part of their daily habits.

It's possible, that I'd add them to an IPFS GUI because I want to back them up, but there is no obvious way (to me) to tell a IPFS node to auto-pin anything that is announced from a specific Peer ID.

This is not a _why add files to ipfs in general_ question. This is, in the context of a desktop / casual user, what is most likely to be my goal. What goals are possible and should we be optimising the flow for?

**Comment by lidel** [comment]
Right now I see two paths that lead to "adding to IPFS":
- user does not really care about data
   (just wants to add it to IPFS to immediately share it with a friend)
   - a replacement for sharing services that expire links after X hours 
- user cares about data
  (keeps it available, shares references to it over longer periods of time)
  - personal semi-public data: websites, notes, photos
  - public data sets (added by IPFS Path)

I am not sure if we want to overthink it. More specific use cases will emerge with time after we provide users with basic UI for (i) adding (ii) managing (iii) sharing files.

**Comment by olizilla** [comment]
Agreed. I wanted to flag the idea that a user adding files to an IPFS app probably doesn't the same goal as a user adding a file to folder in the OS native file browser.

Also agree on not over thinking it. The 2 use-cases you identify are good ones to optimise for right now.
